proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,277 @@
+
+* **Local Caching:** Secrets retrieved successfully from a provider are stored in memory
+* **Background Refresh:** A dedicated background process periodically attempts to refresh each configured remote secret from its provider.
+* **Refresh Interval:** A default refresh interval of 1 hour is used, avoiding excessive calls to external secret providers (typical rotation schedules often measured in hours or days). However, this interval should be configurable on a per-provider instance basis. For example:
+```yaml
+scrape_configs:
+  - job_name: 'example'
+    basic_auth:
+      username: 'user'
+      password:
+        vault: # Example provider
+          address: "https://vault.example.com"
+          path: "secret/data/prometheus/example_job"
+          key: "password"
+          refresh_interval: 30m # Override default for this secret
+    # ... other config ...
+```
+* **Caching Behavior on Failure:** If a scheduled background refresh attempt fails (e.g., due to network issues, temporary provider unavailability, invalid credentials after rotation), Prometheus will continue to use the last successfully fetched secret value. This ensures components relying on the secret can continue operating with the last known good credential, preventing outages due to transient refresh problems. Failed refresh attempts will be logged, and the `prometheus_remote_secret_state` metric will reflect the 'stale' state.
+* **Permission Errors:** If a component using a cached secret reports a specific permission or authentication error, this might indicate the cached secret is no longer valid. Hence, an immediate refresh attempt for that specific secret is triggered, bypassing the regular refresh interval.
+* **Not Querying Per-Scrape:** It is explicitly **not** the default behavior to query the secret provider on every use.
+
+
+### Secret rotation
+
+When a secret is rotated, it is possible for this to happen out of sync. For  example, one of the following could happen with secrets used for target scraping:
+
+1. The scrape target's secret updated but the currently stored secret hasn't updated yet.
+2. The secret provider's secret changes but the scrape target hasn't updated to the new secret yet.
+
+Therefore to help alleviate 1. we can ignore caching and query the secret provider again if permission errors are reported by the component.
+
+To alleviate 2. we can store the previous secret value before rotation for a period of time and use it as a fallback in case of permission errors.
+
+### Metrics
+
+Metrics should be present to help users identify the errors mentioned above (in addition to logs). Therefore the following metrics should be reported per secret present in the config file:
+
+The time since the last successful secret fetch
+
+```
+prometheus_remote_secret_last_successful_fetch_seconds{provider="kubernetes", secret_id="pass1"} 15
+```
+
+A state enum describing in which error condition the secret is in:
```

Comment:
In addition to the state, maybe it would be useful to also expose some counters to show events. Why? because that allow us to identify precisely what is happening across multiple scrapes, network failures etc. Something like:

```
prometheus_remote_secret_fetch_failures_total
prometheus_remote_secret_fetch_success_total
```

Reply:
Agreed & added. Thanks!